File: hier/project-euler/euler-011-hackerrank/euler011.py
#!/usr/bin/env python3

from functools import reduce
from operator import mul

GRID_SIZE = 20  # the measure of each size of the grid
LENGTH = 4      # strand length

grid = [list(map(int, input().split())) for _ in range(GRID_SIZE)]
if any(len(row) != GRID_SIZE for row in grid):
    raise ValueError('wrong row length')

def products():
    for i in range(GRID_SIZE):
        for j in range(GRID_SIZE - LENGTH + 1):
            yield reduce(mul, (grid[i][j + d] for d in range(LENGTH)))

    for i in range(GRID_SIZE - LENGTH + 1):
        for j in range(GRID_SIZE):
            yield reduce(mul, (grid[i + d][j] for d in range(LENGTH)))

    for i in range(GRID_SIZE - LENGTH + 1):
        for j in range(GRID_SIZE - LENGTH + 1):
            yield reduce(mul, (grid[i + d][j + d] for d in range(LENGTH)))

    for i in range(GRID_SIZE - LENGTH + 1):
        for j in range(LENGTH - 1, GRID_SIZE):
            yield reduce(mul, (grid[i + d][j - d] for d in range(LENGTH)))

print(max(products()))

File: hier/project-euler/euler-011-hackerrank/euler011.pl
#!/usr/bin/env perl

use strict;
use warnings;
use Carp qw(croak);
use List::Util qw(all product);

use constant GRID_SIZE => 20;   # the measure of each side of the grid
use constant LEN => 4;          # strand length

my @grid = map { [split q{ }, <>] } (1..GRID_SIZE);
all { @{$_} == GRID_SIZE } @grid or croak 'wrong row length';

my $best = 0;

for my $i (0  .. GRID_SIZE - 1) {
    for my $j (0 .. GRID_SIZE - LEN) {
        my $cur = product map { $grid[$i]->[$j + $_] } (0 .. LEN - 1);
        if ($best < $cur) { $best = $cur }
    }
}

for my $i (0 .. GRID_SIZE - LEN) {
    for my $j (0 .. GRID_SIZE - 1) {
        my $cur = product map { $grid[$i + $_]->[$j] } (0 .. LEN - 1);
        if ($best < $cur) { $best = $cur }
    }
}

for my $i (0 .. GRID_SIZE - LEN) {
    for my $j (0 .. GRID_SIZE - LEN) {
        my $cur = product map { $grid[$i + $_]->[$j + $_] } (0 .. LEN - 1);
        if ($best < $cur) { $best = $cur }
    }
}

for my $i (0 .. GRID_SIZE - LEN) {
    for my $j (LEN - 1 .. GRID_SIZE - 1) {
        my $cur = product map { $grid[$i + $_]->[$j - $_] } (0 .. LEN - 1);
        if ($best < $cur) { $best = $cur }
    }
}

print "$best\n";
